### PR TITLE
API: Add reserved attribute names functionality to the printer.

### PR DESCRIPTION
**Pull request by PapyChacal**
The parser already have an optional parameter on `parse_optional_attr_dict_with_keyword` to handle attribute names that should not be in the generic attribute dictionnary of the custom syntax.
I'd like to do the same thing on the printing side, to simplify func here.

I'm preparing to merge an implementation of the SymbolTable trait, adding a *bunch* of `sym_name` attributes all over, I feel this might help. (e.g. next user is `builtin.module`)

### CODE REVIEW

**Comment by PapyChacal** [comment]
Now, here the custom printer of func still does some extra checks on the reserved keywords, to know if it should print the usual `attributes ` prefix before  the actual dict. Those checks are done by the dict too.
Do we want to add a `prefix` keyword argument to the dictionary printer, to simplify further? I guess that would cover most uses, but I,m concerned about API bloat :thinking:

File: xdsl/printer.py [hunk 1]
```diff
@@ -641,13 +641,17 @@ def _print_attr_string(self, attr_tuple: tuple[str, Attribute]) -> None:
...
 
-    def print_op_attributes(self, attributes: dict[str, Attribute]) -> None:
-        if len(attributes) == 0:
-            return
+    def print_op_attributes(
+        self, attributes: dict[str, Attribute], reserved_attr_names: Iterable[str] = []
+    ) -> None:
+        attribute_list = list(
+            i for i in attributes.items() if i[0] not in reserved_attr_names
+        )
```
Comment (tobiasgrosser) [inline xdsl/printer.py:649]: Is this not just a [filter](https://www.programiz.com/python-programming/methods/built-in/filter).

**Review by tobiasgrosser — APPROVED** [review]
This is a great idea. I guess @math-fehr and maybe @AntonLydike or @superlopuh  might still have feedback.

File: xdsl/printer.py (hunk 1 above)
Comment (math-fehr) [reply xdsl/printer.py:649]: That's the "pythonic" way of writing a filter

File: xdsl/dialects/func.py [hunk 2]
```diff
@@ -177,17 +177,11 @@ def print(self, printer: Printer):
...
-        }
-        if len(attr_dict) > 0:
-            printer.print(" attributes {")
-            printer.print_list(
-                attr_dict.items(), lambda i: printer.print(f'"{i[0]}" = {i[1]}')
-            )
-            printer.print("}")
+
+        reserved = ("sym_name", "function_type", "sym_visibility")
+        if any(k not in reserved for k in self.attributes.keys()):
```
Comment (math-fehr) [inline xdsl/dialects/func.py:182]: Actually, can we add that to `print_op_attributes`?
We probably can do something like adding a `with_keyword` option?
Or adding a new function `print_op_attributes_with_keyword`?
We can do it in another PR if you think this is feature creep ;)

**Review by math-fehr — APPROVED** [review]

File: xdsl/printer.py (hunk 1 above)
Comment (tobiasgrosser) [reply xdsl/printer.py:649]: I was thinking of `list(filter(lambda i: i[0] not in reserved_attr_names, attributes))` or sth similar, but that's just a minor comment.

Comment (superlopuh) [reply xdsl/printer.py:649]: `filter` is significantly slower than list comprehensions, unfortunately

File: xdsl/dialects/func.py [hunk 3]
```diff
@@ -177,17 +177,11 @@ def print(self, printer: Printer):
...
-            if k not in ("sym_name", "function_type", "sym_visibility")
-        }
-        if len(attr_dict) > 0:
-            printer.print(" attributes {")
-            printer.print_list(
-                attr_dict.items(), lambda i: printer.print(f'"{i[0]}" = {i[1]}')
-            )
-            printer.print("}")
+
+        reserved = ("sym_name", "function_type", "sym_visibility")
```
Comment (superlopuh) [inline xdsl/dialects/func.py:181]: create this outside of function, and make a `set`? Seems like good practice even if it won't make a big difference here probably

File: xdsl/printer.py [hunk 4]
```diff
@@ -641,13 +641,17 @@ def _print_attr_string(self, attr_tuple: tuple[str, Attribute]) -> None:
             self.print(f'"{attr_tuple[0]}" = ')
             self.print_attribute(attr_tuple[1])
 
-    def print_op_attributes(self, attributes: dict[str, Attribute]) -> None:
-        if len(attributes) == 0:
-            return
+    def print_op_attributes(
+        self, attributes: dict[str, Attribute], reserved_attr_names: Iterable[str] = []
+    ) -> None:
+        attribute_list = list(
```
Comment (superlopuh) [inline xdsl/printer.py:647]:
```suggestion
        attribute_list = [
```

File: xdsl/printer.py [hunk 5]
```diff
@@ -641,13 +641,17 @@ def _print_attr_string(self, attr_tuple: tuple[str, Attribute]) -> None:
...
+            i for i in attributes.items() if i[0] not in reserved_attr_names
+        )
 
+        if len(attribute_list) == 0:
```
Comment (superlopuh) [inline xdsl/printer.py:651]:
```suggestion
        if not attribute_list:
```

File: xdsl/dialects/func.py (hunk 3 above)
Comment (superlopuh) [reply xdsl/dialects/func.py:181]: This is a nit

**Review by superlopuh — CHANGES_REQUESTED** [review]
LGTM modulo minor changes

File: xdsl/dialects/func.py (hunk 3 above)
Comment (PapyChacal) [reply xdsl/dialects/func.py:181]: I like the idea, making it more declarative, but then it sounds to me like something that could be more integrated to PyRDL?
I'll do the small change, just giving this idea

Comment (superlopuh) [reply xdsl/dialects/func.py:181]: Here I guess it's less about it being declarative and more about Python not being good at constant value optimisation, I think it'll create a tuple every time which seems like a waste if it's constant

File: xdsl/printer.py (hunk 1 above)
Comment (math-fehr) [reply xdsl/printer.py:649]: I wouldn't argue that we should use `filter` because of the speed (unless in the hot section of the code), but rather because of style though.

File: xdsl/dialects/func.py (hunk 3 above)
Comment (math-fehr) [reply xdsl/dialects/func.py:181]: We could have it integrated with `PyRDL`, but we would need to slightly change the parser.
For instance, you could already do `reserved = self.irdl_definition.attributes.keys()`, but that would not be future-proof, as we don't actually check that all attributes are printed, so if you add a new attribute, we might miss it later on.
The declarative framework will ensure that though.

File: xdsl/printer.py (hunk 1 above)
Comment (superlopuh) [reply xdsl/printer.py:649]: They're somewhat related but sure

File: xdsl/dialects/func.py (hunk 2 above)
Comment (PapyChacal) [reply xdsl/dialects/func.py:182]: Yup, I was wondering that! So how do you see this? I can add a `prefix` parameter defaulting to `attributes `?

File: xdsl/printer.py (hunk 4 above)
Comment (PapyChacal) [reply xdsl/printer.py:647]: OOC; is it really different in that case? Aren't both of them calling the list constructor on the same Generator?
No feeling about it, applying your change, but wondering

**Review by superlopuh — APPROVED** [review]

File: xdsl/dialects/func.py (hunk 2 above)
Comment (superlopuh) [reply xdsl/dialects/func.py:181]: I think it makes sense to add this to the same PR tbh

Comment (superlopuh) [reply xdsl/dialects/func.py:181]: Also seems reasonable to add it to the parser also, and add a roundtrip test that the funcitonality between the two functions is somewhat mirrored

**Comment by AntonLydike** [comment]
On a more general level, I am not sure if we should put custom printing syntax into the generic printer. On the purely technical side this looks good!

**Comment by PapyChacal** [comment]
Mh, I get this point.
On the other hand, a lot of custom syntaxes would use such a function, so I just think it makes sense to expose this somewhere; happy to hear ideas on where it would belong!

File: xdsl/dialects/func.py (hunk 2 above)
Comment (math-fehr) [reply xdsl/dialects/func.py:181]: The parser one is already there ^^'

Comment (PapyChacal) [reply xdsl/dialects/func.py:181]: Yup, I wanted to add it as part of this PR and just waited for validation before wasting keystrokes 😋
Will do!

Comment (math-fehr) [reply xdsl/dialects/func.py:181]: SO yeah, I would just put a boolean rather than a string. I don't think anyone would use a different keyword than `attributes `, and we probably don't want to give the option to people (otherwise our code will become less easy to read).

**Review by superlopuh — APPROVED** [review]

**Comment by math-fehr** [comment]
Looks good to me! I would just remove the `with_keyword` in the name though, and that's good to merge for me!

File: xdsl/printer.py [hunk 6]
```diff
@@ -652,6 +652,26 @@ def print_op_attributes(self, attributes: dict[str, Attribute]) -> None:
 
         self.print("}")
 
+    def print_op_attributes_with_keyword(
```
Comment (superlopuh) [inline xdsl/printer.py:655]: It feels like this could be fused with the previous method, with a pretty low overhead. I'd just check whether reserved attr names is not empty before filtering the attribute list

Comment (superlopuh) [reply xdsl/printer.py:655]: Oh and please remove the `if len(attributes) == 0` and replace with `if not attributes`

**Comment by PapyChacal** [comment]
Discussed API naming issues in private, agreed on what I just pushed, there was just incompatible requests here.
NB: Agreed that it isn't the best naming, but I was just after factoring this first. Let's rework this API whenever useful!
